src/commons/s3.rs

```diff
@@ -0,0 +1,206 @@
+#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum S3BucketDef {
+    Inline(S3BucketSpec),
+    Reference(String),
+}
+
+impl S3BucketDef {
+    /// Returns an [InlinedS3BucketSpec].
+    pub async fn resolve(
+        &self,
+        client: &Client,
+        namespace: Option<String>,
+    ) -> OperatorResult<InlinedS3BucketSpec> {
+        match self {
+            S3BucketDef::Inline(s3_bucket) => s3_bucket.inlined(client, namespace).await,
+            S3BucketDef::Reference(_s3_bucket) => {
+                S3BucketSpec::get(_s3_bucket.as_str(), client, namespace.clone())
+                    .await?
+                    .inlined(client, namespace)
+                    .await
+            }
+        }
+    }
+}
+
+/// S3 connection definition used by [S3BucketSpec]
+#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum ConnectionDef {
+    Inline(S3ConnectionSpec),
+    Reference(String),
+}
+
+/// S3 connection definition as CRD.
+#[derive(CustomResource, Clone, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[kube(
+    group = "s3.stackable.tech",
+    version = "v1alpha1",
+    kind = "S3Connection",
+    plural = "s3connections",
+    crates(
+        kube_core = "kube::core",
+        k8s_openapi = "k8s_openapi",
+        schemars = "schemars"
+    ),
+    namespaced
+)]
+#[serde(rename_all = "camelCase")]
+pub struct S3ConnectionSpec {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub host: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub port: Option<u16>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub secret_class: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub tls: Option<Tls>,
+}
+impl S3ConnectionSpec {
+    /// Convenience function to retrieve the spec of a S3 connection resource from the K8S API service.
+    pub async fn get(
+        resource_name: &str,
+        client: &Client,
+        namespace: Option<String>,
+    ) -> OperatorResult<S3ConnectionSpec> {
+        client
+            .get::<S3Connection>(resource_name, namespace.as_deref())
+            .await
+            .map(|conn| conn.spec)
+            .map_err(|_source| error::Error::MissingS3Connection {
+                name: resource_name.to_string(),
+            })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::commons::s3::ConnectionDef::Inline;
+    use crate::commons::s3::{S3BucketSpec, S3ConnectionSpec};
+
+    #[test]
+    fn test_ser_inline() {
+        let bucket = S3BucketSpec {
+            bucket_name: Some("test-bucket-name".to_owned()),
+            connection: Some(Inline(S3ConnectionSpec {
+                host: Some("host".to_owned()),
+                port: Some(8080),
+                secret_class: None,
+                tls: None,
+            })),
+        };
```

Comment:
```suggestion
    use crate::commons::s3::ConnectionDef;
    use crate::commons::s3::{S3BucketSpec, S3ConnectionSpec};

    #[test]
    fn test_ser_inline() {
        let bucket = S3BucketSpec {
            bucket_name: Some("test-bucket-name".to_owned()),
            connection: Some(ConnectionDef::Inline(S3ConnectionSpec {
                host: Some("host".to_owned()),
                port: Some(8080),
                secret_class: None,
                tls: None,
            })),
        };
```
Just a test and just a preference, you can ignore this if you do not like it :)